### ISSUE TITLE
Conch ssh client auth order doesn't honour preferredOrder

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4266]
|<img alt="alanfranzoni's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| alanfranzoni reported|
|-|-|
|Trac ID|trac#4266|
|Type|defect|
|Created|2010-02-10 00:09:47Z|
|Branch|https://github.com/twisted/twisted/tree/userauth-order-4266|

Problem lies in twisted.conch.ssh.userauth.py, class SSHUserAuthClient

when trying to auth, ordering doesn't work as expected; if preferredorder is ["a", "b"] and server return as auth methods ["x" , "a"], the client will try to use, in order, ["x", "a"]. This happens because orderByPreference in ssh_USERAUTH_FAILURE() returns -1 when a method is NOT within preferredOrder, and util.dsu will then put this method to the beginning of the resulting canContinue list.

The test wasn't really suited to check for that error because the auth method "afirstmethod" did not exist on the client, and this just made the client skip that method because it couldn't fine auth_afirstmethod, and go on with the password.

The attached patch contains a modifed test which triggers the issue and a proposed fix.

Attachments:

* [twisted_ssh_userauth.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5ed/5ed1cabe0624db28b5d4f9aa9929542455d5264c/9f5d4bdcb89a65752387f13cec00482939fd7650.patch) (1438 bytes) - added by alanfranzoni on 2010-02-10 00:10:20Z - ssh client userauth correct preferredOrder patch
* [userauth-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5ed/5ed1cabe0624db28b5d4f9aa9929542455d5264c/e36cf8652a6b0e0a403737068edcf72dc410dc9f.patch) (2292 bytes) - added by alanfranzoni on 2010-02-10 20:44:52Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4266 4266
type__defect defect
reporter__alanfranzoni alanfranzoni
priority__normal normal
milestone__ 
branch__branches_userauth_order_4266 branches/userauth-order-4266
branch_author__jesstess jesstess
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1265760587000000 1265760587000000
changetime__1265922483000000 1265922483000000
version__None None
owner__ 
cc__jesstess cc__alanfranzoni
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_3"></a>|@jesstess commented|
|-|-|

(In [28160]) Branching to 'userauth-order-4266'

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_4"></a>|@jesstess commented|
|-|-|

(In [28161]) Apply twisted_ssh_userauth.patch by alanfranzoni.

refs [#4266](https://github.com/twisted/twisted/issues/4266)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_5"></a>|@jesstess set owner to alanfranzoni|
|-|-|

Thanks for the patch, alanfranzoni. The patch looks good and the branch passes on my test machine. A few comments:

 * orderByPreference would benefit from a doc string, including epytext markup for it's parameter, and ssh_USERAUTH_FAILURE could also use markup for it's parameters and return value. Docstrings and markup are described in the [coding standard](http://twistedmatrix.com/trac/browser/trunk/doc/core/development/policy/coding-standard.xhtml).
 * Since userauth.py and test_userauth.py are getting modified, can you bump the copyrights on them?
 * Can you add a NEWS file for the defect, as described in ReviewProcess?

Branch userauth-order-4266 now exists for this change. Please make subsequent diffs against this branch. Thanks again!

**Comment by twisted-trac** [comment]
|<img alt="alanfranzoni's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|alanfranzoni set owner to @jesstess|
|-|-|

Here it is, I hope I have understood the way the conch client works :-)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_8"></a>|@jesstess commented|
|-|-|

(In [28170]) Apply userauth-2.patch and add a bit up epytext markup.

refs [#4266](https://github.com/twisted/twisted/issues/4266)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_10"></a>|@jesstess set status to `closed`|
|-|-|

(In [28171]) Merge userauth-order-4266

Author: alanfranzoni
Reviewer: jesstess
Fixes: [#4266](https://github.com/twisted/twisted/issues/4266)

twisted.conch.ssh.SSHUserAuthClient now honors preferredOrder when
authenticating.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
